ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/AttestationManager.java

```diff
@@ -49,38 +52,80 @@
+
   AttestationManager(
       final EventBus eventBus,
       final ForkChoice attestationProcessor,
       final PendingPool<ValidateableAttestation> pendingAttestations,
       final FutureItems<ValidateableAttestation> futureAttestations,
-      final AggregatingAttestationPool aggregatingAttestationPool) {
+      final AggregatingAttestationPool aggregatingAttestationPool,
+      final AttestationValidator attestationValidator,
+      final AggregateAttestationValidator aggregateValidator) {
     this.eventBus = eventBus;
     this.attestationProcessor = attestationProcessor;
     this.pendingAttestations = pendingAttestations;
     this.futureAttestations = futureAttestations;
     this.aggregatingAttestationPool = aggregatingAttestationPool;
+    this.attestationValidator = attestationValidator;
+    this.aggregateValidator = aggregateValidator;
   }
 
   public static AttestationManager create(
       final EventBus eventBus,
       final PendingPool<ValidateableAttestation> pendingAttestations,
       final FutureItems<ValidateableAttestation> futureAttestations,
       final ForkChoice attestationProcessor,
-      final AggregatingAttestationPool aggregatingAttestationPool) {
+      final AggregatingAttestationPool aggregatingAttestationPool,
+      final AttestationValidator attestationValidator,
+      final AggregateAttestationValidator aggregateValidator) {
     return new AttestationManager(
         eventBus,
         attestationProcessor,
         pendingAttestations,
         futureAttestations,
-        aggregatingAttestationPool);
+        aggregatingAttestationPool,
+        attestationValidator,
+        aggregateValidator);
   }
 
   public void subscribeToProcessedAttestations(
       ProcessedAttestationListener processedAttestationListener) {
     processedAttestationSubscriber.subscribe(processedAttestationListener);
   }
 
+  public SafeFuture<InternalValidationResult> addAttestation(ValidateableAttestation attestation) {
+    SafeFuture<InternalValidationResult> validationResult =
+        attestationValidator.validate(attestation);
+    processInternallyValidatedAttestation(validationResult, attestation);
+    return validationResult;
+  }
+
+  public SafeFuture<InternalValidationResult> addAggregate(ValidateableAttestation attestation) {
+    SafeFuture<InternalValidationResult> validationResult =
+        aggregateValidator.validate(attestation);
+    processInternallyValidatedAttestation(validationResult, attestation);
+    return validationResult;
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void processInternallyValidatedAttestation(
+      SafeFuture<InternalValidationResult> validationResult, ValidateableAttestation attestation) {
+    validationResult.thenAccept(
+        internalValidationResult -> {
+          if (internalValidationResult.equals(InternalValidationResult.ACCEPT)
+              || internalValidationResult.equals(InternalValidationResult.SAVE_FOR_FUTURE)) {
```

Comment:
If we already know the attestation should be saved for future, do we need to run it through the attestation processor or can it just be added to the pending pool immediately?

Reply:
Made an issue here to solve later on: https://github.com/ConsenSys/teku/issues/3114